data/cg-teams.ts

```diff
@@ -376,96 +459,137 @@ export default class TeamGenerator {
 		if (move.flags.punch && ability === 'Iron Fist') weight *= 1.2;
-		if (!move.flags.protect) weight *= 1.1;
+		if (!move.flags.protect) weight *= 1.05;
 		if (move.flags.slicing && ability === 'Sharpness') weight *= 1.5;
+		if (move.flags.sound && ability === 'Punk Rock') weight *= 1.3;
 
 		// boosts/secondaries
 		// TODO: consider more possible secondaries
-		weight *= this.boostWeight(move, movesSoFar, species);
-		if (move.secondary?.status) {
-			weight *= TeamGenerator.statusWeight(move.secondary.status, (move.secondary.chance || 100) / 100);
+		weight *= this.boostWeight(move, movesSoFar, species, ability, level);
+		const secondaryChance = Math.min((move.secondary?.chance || 100) * (ability === 'Serene Grace' ? 2 : 1) / 100, 100);
+		if (move.secondary?.status && ability !== 'Sheer Force') {
+			weight *= TeamGenerator.statusWeight(move.secondary.status, secondaryChance);
 		}
+		if (ability === 'Toxic Chain') weight *= TeamGenerator.statusWeight('tox', 1 - Math.pow(0.7, numberOfHits));
 
 		// self-inflicted confusion or locking yourself in
 		if (move.self?.volatileStatus) weight *= 0.8;
 
 		// downweight moves if we already have an attacking move of the same type
-		if (movesSoFar.some(m => m.category !== 'Status' && m.type === move.type && m.basePower >= 60)) weight *= 0.3;
+		if (movesSoFar.some(m => m.category !== 'Status' && m.type === moveType && m.basePower >= 60)) weight *= 0.3;
 
 		if (move.selfdestruct) weight *= 0.3;
-		if (move.recoil) weight *= 1 - (move.recoil[0] / move.recoil[1]);
+		if (move.recoil && ability !== 'Rock Head' && ability !== 'Magic Guard') {
+			weight *= 1 - (move.recoil[0] / move.recoil[1]);
+			if (ability === 'Reckless') weight *= 1.2;
+		}
 		if (move.mindBlownRecoil) weight *= 0.25;
 		if (move.flags['futuremove']) weight *= 0.3;
-		// TODO: account for normal higher-crit-chance moves
-		if (move.willCrit) weight *= 1.45;
+
+		let critRate = move.willCrit ? 4 : move.critRatio || 1;
+		if (ability === 'Super Luck') critRate++;
+		if (movesSoFar.some(m => m.id === 'focusenergy')) {
+			critRate += 2;
+			weight *= 0.9; // a penalty the extra turn of setup
+		}
+		if (critRate > 4) critRate = 4;
+		weight *= 1 + [0, 1 / 24, 1 / 8, 1 / 2, 1][critRate] * (ability === 'Sniper' ? 1 : 0.5);
+
+		// these two hazard removers don't clear hazards on the opponent's field, but can be blocked by type immunities
+		if (['rapidspin', 'mortalspin'].includes(move.id)) {
+			weight *= 1 + 20 * Math.pow(0.25, teamStats.hazardRemovers);
+			teamStats.hazardRemovers++;
+		}
 
 		if (move.drain) {
 			const drainedFraction = move.drain[0] / move.drain[1];
 			weight *= 1 + (drainedFraction * 0.5);
 		}
 
-		// don't need 2 healing moves
-		if (move.heal && movesSoFar.some(m => m.heal)) weight *= 0.5;
+		// Oricorio should rarely get Tera Blast, as Revelation Dance is strictly better
+		if (species.baseSpecies === 'Oricorio' && move.id === 'terablast') weight *= 0.5;
```

Comment:
I'm worried that these sorts of hardcodes will grow and make this format more unwieldy and similar to randbats... but also I don't have the time to maintain this format right now, so I suppose this one time it's OK.